dataproc: add CreateCluster and Quickstart samples

bradmiro opened the pull request:
Adding two samples for Dataproc: CreateCluster, and Quickstart

krousey commented on dataproc/dataproc_quickstart/quickstart.go:
```diff
@@ -0,0 +1,194 @@
+// Cloud Dataproc cluster
+
+// [START dataproc_quickstart]
+package main
```
Does this compile without a `func main()`?

I think the test will run because the testing tool supplies it's own `func main()`, but does `go build` work?

bradmiro replied:
Good catch, it does not! Looking into this...

tbpg replied:
Sorry that's not clear. Yes, please switch this to `func main` with flags.

krousey commented on dataproc/dataproc_quickstart/quickstart.go:
```diff
@@ -0,0 +1,194 @@
+	start := time.Now().UTC()
+
+	state := submitJobResp.Status.State
+	for _, ok := ts[state]; !ok; _, ok = ts[state] {
```
This could be simplified to
```go
for !ts[state] {
```

since a map lookup on a missing key returns the default value of `false` here.

bradmiro replied:
Agreed and fixed, thanks

krousey commented on dataproc/dataproc_quickstart/quickstart.go:
```diff
@@ -0,0 +1,194 @@
+	endpoint := fmt.Sprintf("%s-dataproc.googleapis.com:443", region)
+	clusterClient, err := dataproc.NewClusterControllerClient(ctx, option.WithEndpoint(endpoint))
+	if err != nil {
+		return fmt.Errorf("Error creating the cluster client: %s", err)
```
Minor nit: Generally you don't capitalize error messages since they tend to be concatenated and/or wrapped in other error messages.

bradmiro replied:
Fixed, thanks

tbpg commented on dataproc/create_cluster/create_cluster.go:
```diff
@@ -0,0 +1,77 @@
+// Copyright 2019 Google LLC
```
Filepath comment: looks like this is the first Dataproc sample. Could we put all of the snippets into the top-level `dataproc` directory and put the quickstart in a `dataproc/quickstart` directory? No need to duplicate `dataproc` in the name.

Also, if there will be additional cluster samples in the future (create, delete, list, etc.), a pattern I've liked for other samples is starting the filename with the name of the thing -- `cluster_create.go`, `cluster_delete.go`, etc. Not mandatory at all, but can make it easier to find the right file for each sample.

bradmiro replied:
Moved create_cluster to top-level but leaving file names as-is to match method names / file names in other languages.

tbpg commented on dataproc/create_cluster/create_cluster.go:
```diff
@@ -0,0 +1,77 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// This sample shows how you can use the Cloud Dataproc Client library to create a
```
```suggestion
// The dataproc package shows how you can use the Cloud Dataproc Client library.
```

Also, no newline after the comment. https://github.com/golang/go/wiki/CodeReviewComments#package-comments

bradmiro replied:
After changing this, the linter suggested `// Package dataproc shows how...`, what do you think?

tbpg commented on dataproc/create_cluster/create_cluster.go:
```diff
@@ -0,0 +1,77 @@
+	ctx := context.Background()
+
+	clusterClient, err := dataproc.NewClusterControllerClient(ctx,
+		option.WithEndpoint(region+"-dataproc.googleapis.com:443"),
```
A single line is probably OK. In general, Go doesn't have a line length limit. In particular, discourages line breaks in the middle of function calls. See https://github.com/golang/go/wiki/CodeReviewComments#line-length.

However, sometimes it does make sense to have a line break to avoid having horizontal scroll when the sample is embedded in docs. In this case, it's probably short enough to have on a single line.

tbpg commented on dataproc/create_cluster/create_cluster.go:
```diff
@@ -0,0 +1,77 @@
+	dataprocpb "google.golang.org/genproto/googleapis/cloud/dataproc/v1"
+)
+
+func createCluster(w io.Writer, projectID, region, clusterName string) error {
```
Please include sample values. https://github.com/GoogleCloudPlatform/golang-samples/blob/master/CONTRIBUTING.md#function-arguments

tbpg commented on dataproc/create_cluster/create_cluster.go:
```diff
@@ -0,0 +1,77 @@
+		return fmt.Errorf("Error creating the cluster client: %s", err)
+	}
+
+	// Create the cluster config
```
Nit: period at the end. Same with the other comments.

tbpg commented on dataproc/create_cluster/create_cluster.go:
```diff
@@ -0,0 +1,77 @@
+	// Create the cluster
+	op, err := clusterClient.CreateCluster(ctx, req)
+	if err != nil {
+		return fmt.Errorf("Error submitting the cluster creation request: %s", err)
```
Can shorten this to:
```suggestion
		return fmt.Errorf("CreateCluster: %v", err)
```

(Also `%v` instead of `%s` for errors.)

Same with other errors. https://github.com/GoogleCloudPlatform/golang-samples/blob/master/CONTRIBUTING.md#return-errors

tbpg commented on dataproc/create_cluster/create_cluster_test.go:
```diff
@@ -0,0 +1,76 @@
+)
+
+func setClusterName(tc testutil.Context) {
+	clusterName = tc.ProjectID + "-test-dataproc-create-cluster"
```
Please inline this and avoid the global var.

tbpg commented on dataproc/create_cluster/create_cluster_test.go:
```diff
@@ -0,0 +1,76 @@
+)
+
+var (
+	region      = "us-central1"
```
```suggestion
	const region      = "us-central1"
```

(Not a var.)

Alternatively, inline in the test, since it's not used anywhere else.

tbpg commented on dataproc/create_cluster/create_cluster_test.go:
```diff
@@ -0,0 +1,76 @@
+	clusterName = tc.ProjectID + "-test-dataproc-create-cluster"
+}
+
+func teardown(t *testing.T, tc testutil.Context) {
```
Please add a call to `t.Helper()`. That makes it so if this fails, the error shows the line where this function was called, rather than the line within this function that failed. That makes it easier to track down what test failed. https://golang.org/pkg/testing/#T.Helper

tbpg commented on dataproc/create_cluster/create_cluster_test.go:
```diff
@@ -0,0 +1,76 @@
+	a := fmt.Sprintf("%s-dataproc.googleapis.com:443", region)
+	clusterClient, err := dataproc.NewClusterControllerClient(ctx, option.WithEndpoint(a))
+
+	clusterName := clusterName
```
Make an arg. Also, no need for a separate var within this function.

tbpg commented on dataproc/create_cluster/create_cluster_test.go:
```diff
@@ -0,0 +1,76 @@
+
+	got := buf.String()
+	if want := fmt.Sprintf("successfully: %s", clusterName); !strings.Contains(got, want) {
+		t.Fatalf("got %s, want %s", got, want)
```
Please include the name of the function that failed in the message. https://github.com/golang/go/wiki/CodeReviewComments#useful-test-failures

tbpg commented on dataproc/create_cluster/create_cluster_test.go:
```diff
@@ -0,0 +1,76 @@
+func teardown(t *testing.T, tc testutil.Context) {
+	ctx := context.Background()
+
+	a := fmt.Sprintf("%s-dataproc.googleapis.com:443", region)
```
What does `a` stand for? Perhaps change to `endpoint`?

bradmiro replied:
I moved `a` to just be inline, per one of your previous comments

tbpg commented on dataproc/dataproc_quickstart/quickstart.go:
```diff
@@ -0,0 +1,194 @@
+		// fmt.Fprintf(w, "-----\n")
+		// fmt.Fprintf(w, "State: %s", state)
+		// fmt.Fprintf(w, "out: %t, ok: %t", out, ok)
+		// fmt.Fprintf(w, "-----\n")
```
Delete?

tbpg commented on dataproc/dataproc_quickstart/quickstart.go:
```diff
@@ -0,0 +1,194 @@
+	fmt.Fprintf(w, "Submitted job %q", id)
+
+	// These states all signify that a job has terminated, successfully or not
+	ts := map[dataprocpb.JobStatus_State]bool{
```
Might do a longer variable name since `ts` isn't clear. `terminalState` or `terminalStates`?

tbpg commented on dataproc/dataproc_quickstart/quickstart.go:
```diff
@@ -0,0 +1,194 @@
+	timeout := 5 * time.Minute
+	start := time.Now().UTC()
+
+	state := submitJobResp.Status.State
```
Consider checking this inside the for loop and calling `break` when you reach the stopping condition.

tbpg commented on dataproc/dataproc_quickstart/quickstart.go:
```diff
@@ -0,0 +1,194 @@
+		// fmt.Fprintf(w, "-----\n")
+		if time.Now().UTC().Sub(start) > timeout {
+			cancelReq := &dataprocpb.CancelJobRequest{ProjectId: projectID, Region: region, JobId: id}
+			jobClient.CancelJob(ctx, cancelReq)
```
Please check if this returns an error.

bradmiro replied:
Fixed

tbpg commented on dataproc/dataproc_quickstart/quickstart.go:
```diff
@@ -0,0 +1,194 @@
+			return fmt.Errorf("Job %q timed out after threshold of %d seconds", id, int64(timeout.Seconds()))
+		}
+
+		time.Sleep(1 * time.Second)
```
Sleep at the end?

tbpg commented on dataproc/dataproc_quickstart/quickstart.go:
```diff
@@ -0,0 +1,194 @@
+		// fmt.Fprintf(w, "out: %t, ok: %t", out, ok)
+		// fmt.Fprintf(w, "-----\n")
+		if time.Now().UTC().Sub(start) > timeout {
+			cancelReq := &dataprocpb.CancelJobRequest{ProjectId: projectID, Region: region, JobId: id}
```
Should probably be consistent with doing one line or multiple lines for requests. I'd probably do multiple.

tbpg commented on dataproc/dataproc_quickstart/quickstart.go:
```diff
@@ -0,0 +1,194 @@
+
+	// We can create a timeout such that the job gets cancelled if not in a terminal state after a certain amount of time
+	timeout := 5 * time.Minute
+	start := time.Now().UTC()
```
I'm not sure this is the best way to do this. Will update this comment with a suggestion after this initial review.

bradmiro replied:
Would definitely appreciate some help here :D there is no way to naturally handle this with the Dataproc API (yet...)

bradmiro replied:
I slightly improved this and I think this works better

tbpg commented on dataproc/dataproc_quickstart/quickstart.go:
```diff
@@ -0,0 +1,194 @@
+		return fmt.Errorf("Error getting cluster %q with error: %v", clusterName, err)
+	}
+
+	sc, err := storage.NewClient(ctx)
```
Probably `storageClient`. Consistent client naming.

tbpg commented on dataproc/dataproc_quickstart/quickstart.go:
```diff
@@ -0,0 +1,194 @@
+	}
+
+	obj := fmt.Sprintf("google-cloud-dataproc-metainfo/%s/jobs/%s/driveroutput.000000000", resp.ClusterUuid, id)
+	rc, err := sc.Bucket(resp.Config.ConfigBucket).Object(obj).NewReader(ctx)
```
`rc`?

tbpg commented on dataproc/dataproc_quickstart/quickstart.go:
```diff
@@ -0,0 +1,194 @@
+
+	defer rc.Close()
+
+	body, err := ioutil.ReadAll(rc)
```
Is it OK to assume this can fit in memory?

bradmiro replied:
While technically there could be an obscure edge-case, I don't feel it necessary to account for this.

bradmiro replied:
Confirming let's leave this as-is

tbpg commented on dataproc/dataproc_quickstart/quickstart.go:
```diff
@@ -0,0 +1,194 @@
+	fmt.Fprintf(w, "Job %q finished with state %s:\n%s", id, state, body)
+
+	// Delete the cluster once the job has terminated
+	dReq := &dataprocpb.DeleteClusterRequest{
```
Should this be in a `defer` after the cluster is created to make sure the cluster gets cleaned up even if there is a return above? Need to be careful calling `log.Fatal`, though, since `defer`s don't run when you call `os.Exit`, which `log.Fatal` does.

bradmiro replied:
Is there a terminal `error` message that triggers `defer`?

bradmiro replied:
`panic(fmt.Sprintf("error: %v", err))`?

bradmiro replied:
Actually, I want to check with the Dataproc folks about this.

bradmiro commented:
@tbpg implemented many of your changes and added testing support to quickstart.go. I also added region tags such that the `flag` elements will be hidden on cloud.google.com.

tbpg commented on dataproc/create_cluster.go:
```diff
@@ -0,0 +1,77 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package dataproc package shows how you can use the Cloud Dataproc Client library to create a
```
```suggestion
// Package dataproc shows how you can use the Cloud Dataproc Client library to create a
```

tbpg commented on dataproc/create_cluster.go:
```diff
@@ -0,0 +1,77 @@
+)
+
+func createCluster(w io.Writer, projectID, region, clusterName string) error {
+	// projectID = "your-project-id"
```
```suggestion
	// projectID := "your-project-id"
```

Same for the others.

tbpg commented on dataproc/create_cluster.go:
```diff
@@ -0,0 +1,77 @@
+	ctx := context.Background()
+
+	// Create the cluster client.
+	clusterClient, err := dataproc.NewClusterControllerClient(ctx, option.WithEndpoint(region + "-dataproc.googleapis.com:443"))
```
Optional. The line is slightly long, so it might be good to create another variable? Would match the quickstart.

```suggestion
endpoint := region + "-dataproc.googleapis.com:443"
clusterClient, err := dataproc.NewClusterControllerClient(ctx, option.WithEndpoint(endpoint))
```

tbpg commented on dataproc/create_cluster.go:
```diff
@@ -0,0 +1,77 @@
+	// Create the cluster client.
+	clusterClient, err := dataproc.NewClusterControllerClient(ctx, option.WithEndpoint(region + "-dataproc.googleapis.com:443"))
+	if err != nil {
+		return fmt.Errorf("CreateCluster: %v", err)
```
```suggestion
		return fmt.Errorf("dataproc.NewClusterControllerClient: %v", err)
```

tbpg commented on dataproc/create_cluster.go:
```diff
@@ -0,0 +1,77 @@
+	fmt.Fprintf(w, "Cluster created successfully: %s", resp.ClusterName)
+	return nil
+}
+// [END dataproc_create_cluster]
```
Please add a newline at the end (I think `gofmt` does this automatically).

tbpg commented on dataproc/quickstart/quickstart.go:
```diff
@@ -0,0 +1,221 @@
+	// clusterName = "your-cluster-name"
+	// jobFilePath = "gs://your_file/location"
+	// [END dataproc_quickstart]
+	flag.Parse()
```
I think it's OK to show the flag parsing in the sample. It's what we do for other quickstarts. Maybe they don't all need to be flags? Could add comments saying to update as needed? The project ID needs to be a flag to test, though.

bradmiro replied:
For proper testing I believe these are all necessary

tbpg commented on dataproc/quickstart/quickstart.go:
```diff
@@ -0,0 +1,221 @@
+
+			_, err := jobClient.CancelJob(ctx, cancelReq)
+			if err != nil {
+				fmt.Errorf("error cancelling job: %v", err)
```
Should these be `log.Fatalf`? Or `log.Printf` and break?

tbpg commented on dataproc/quickstart/quickstart.go:
```diff
@@ -0,0 +1,221 @@
+			}
+
+			_, err := jobClient.CancelJob(ctx, cancelReq)
+			if err != nil {
```
```suggestion
			if _, err := jobClient.CancelJob(ctx, cancelReq); err != nil {
```

tbpg commented on dataproc/quickstart/quickstart.go:
```diff
@@ -0,0 +1,221 @@
+			fmt.Errorf("error getting job %q with error: %v", id, err)
+			break
+		}
+		state = getJobResp.Status.State
```
Consider putting the state check here and breaking on a terminal state. That way, you don't have to look up to the top of the loop to find when it stops (or keep that info in your head while reading the body of the loop). That also has the benefit of not sleeping after the job is done.

tbpg commented on dataproc/quickstart/quickstart.go:
```diff
@@ -0,0 +1,221 @@
+		}
+		state = getJobResp.Status.State
+
+		time.Sleep(1 * time.Second)
```
Any particular reason for 1 second? Also, add a comment saying why this sleeps?

bradmiro replied:
This is to prevent excess polling of the api. The value is fairly arbitrary but I've been using 1 second for all of the quickstarts

tbpg commented on dataproc/quickstart/quickstart.go:
```diff
@@ -0,0 +1,221 @@
+	}
+
+	obj := fmt.Sprintf("google-cloud-dataproc-metainfo/%s/jobs/%s/driveroutput.000000000", resp.ClusterUuid, id)
+	reader, err := storageClient.Bucket(resp.Config.ConfigBucket).Object(obj).NewReader(ctx)
```
Should we only read the result if the job does not time out? Use `defer` for the delete so we can `return` on timeout and the cluster will still get deleted? Or maybe put it into a function you can call here and in the cancel block?

tbpg commented on dataproc/quickstart/quickstart_test.go:
```diff
@@ -0,0 +1,159 @@
+
+// Underscored names so as to not collide with sample declarations
+var (
+	_clusterName string
```
No underscores in Go variable names. Just clusterName (or a different name, if it should be). https://golang.org/doc/effective_go.html#mixed-caps

tbpg commented on dataproc/quickstart/quickstart_test.go:
```diff
@@ -0,0 +1,159 @@
+	}
+
+	bkt := sc.Bucket(bktName)
+	if err := bkt.Create(ctx, tc.ProjectID, nil); err != nil {
```
Tests get messed up and buckets don't get deleted. Please copy and paste https://github.com/GoogleCloudPlatform/golang-samples/blob/master/storage/objects/main_test.go#L363. (I'll move that to a shared location in a future PR.)

tbpg commented on dataproc/quickstart/quickstart_test.go:
```diff
@@ -0,0 +1,159 @@
+		t.Errorf("Error deleting bucket: %v", err)
+	}
+
+	ep := fmt.Sprintf("%s-dataproc.googleapis.com:443", _region)
```
`endpoint` (same as sample code).

tbpg commented on dataproc/quickstart/quickstart_test.go:
```diff
@@ -0,0 +1,159 @@
+		"--project_id", tc.ProjectID,
+		"--region", _region,
+		"--cluster_name", _clusterName,
+        "--job_file_path", _jobFilePath,
```
Need to run `gofmt -w -s .`.

tbpg commented on dataproc/quickstart/quickstart.go:
```diff
@@ -0,0 +1,221 @@
+// Cloud Dataproc cluster, submit a PySpark job to the cluster, wait for the job to finish
+// and finally delete the cluster.
+
+// [START dataproc_quickstart]
```
```
// [START dataproc_quickstart]

// The quickstart command shows how you can...
package main
```

Nice to have the comment in the sample. The newline between the region tag and the comment makes it so the region tag doesn't show up in the GoDoc for the command.

tbpg commented on dataproc/quickstart/quickstart.go:
```diff
@@ -181,41 +191,27 @@ func main() {
 	resp, err := clusterClient.GetCluster(ctx, getCReq)
 	if err != nil {
-		fmt.Errorf("error getting cluster %q: %v", clusterName, err)
+		fmt.Println(fmt.Errorf("error getting cluster %q: %v", clusterName, err))
```
No need to create an error then print it. Can just print directly. Return after this (and the other errors)?

bradmiro commented on dataproc/quickstart/quickstart.go:
```diff
@@ -140,21 +147,20 @@ func main() {
 
-	state := submitJobResp.Status.State
-	for !terminalStates[state] {
+	var state dataprocpb.JobStatus_State
```
`state` has usage outside the scope of the loop as I include in the final output with the job output
